### PR TITLE
Bug ID: #2839

### PR DESCRIPTION
**Pull request by rpsec**
I added an additional function to only return the relative path of an index file. This worked for some basic tests for adding/updating/searching a repo. Still needs some additional testing to make sure the %s-index.yaml files are being read properly.

Fixes: https://github.com/kubernetes/helm/issues/2839

### CODE REVIEW

**Review by bacongobbler — CHANGES_REQUESTED** [review]
would you mind writing some unit tests as well? Thanks!

**Comment by rprobotics** [comment]
Got it! I can get started on that today, and hopefully have something pushed sometime this week.

**Comment by rprobotics** [comment]
My unit tests have been added, and passed locally.

**Comment by bacongobbler** [comment]
Here's all the references to `CacheIndex` that may or may not need to be updated similar to `repo add`:

```
><> git grep CacheIndex
cmd/helm/dependency_build_test.go:      i, err := repo.LoadIndexFile(dbc.helmhome.CacheIndex("test"))
cmd/helm/dependency_update_test.go:     i, err := repo.LoadIndexFile(duc.helmhome.CacheIndex("test"))
cmd/helm/helm_test.go:          os.Symlink(localRepoIndexFile, home.CacheIndex("local"))
cmd/helm/home.go:                               fmt.Fprintf(out, "Stable CacheIndex: %s\n", h.CacheIndex("stable"))
cmd/helm/init.go:               sr, err := initStableRepo(home.CacheIndex(stableRepository), skipRefresh)
cmd/helm/init.go:               lr, err := initLocalRepo(home.LocalRepository(localRepositoryIndexFile), home.CacheIndex("local"))
cmd/helm/repo_remove.go:        if _, err := os.Stat(home.CacheIndex(name)); err == nil {
cmd/helm/repo_remove.go:                err = os.Remove(home.CacheIndex(name))
cmd/helm/repo_remove_test.go:   mf, _ := os.Create(hh.CacheIndex(testName))
cmd/helm/repo_remove_test.go:   if _, err := os.Stat(hh.CacheIndex(testName)); err == nil {
cmd/helm/repo_update_test.go:           Cache: hh.CacheIndex("charts"),
cmd/helm/search.go:             f := s.helmhome.CacheIndex(n)
pkg/downloader/chart_downloader.go:     i, err := repo.LoadIndexFile(c.HelmHome.CacheIndex(r.Config.Name))
pkg/downloader/chart_downloader.go:             i, err := repo.LoadIndexFile(c.HelmHome.CacheIndex(r.Config.Name))
pkg/downloader/manager.go:              cacheindex := m.HelmHome.CacheIndex(lname)
pkg/helm/helmpath/helmhome.go:// CacheIndex returns the path to an index for the given named repository.
pkg/helm/helmpath/helmhome.go:func (h Home) CacheIndex(name string) string {
pkg/helm/helmpath/helmhome_unix_test.go:        isEq(t, hh.CacheIndex("t"), "/r/repository/cache/t-index.yaml")
pkg/helm/helmpath/helmhome_windows_test.go:     isEq(t, hh.CacheIndex("t"), "r:\\repository\\cache\\t-index.yaml")
pkg/repo/repotest/server.go:            Cache: home.CacheIndex(name),
pkg/resolver/resolver.go:               repoIndex, err := repo.LoadIndexFile(r.helmhome.CacheIndex(repoNames[d.Name]))
```

**Comment by rprobotics** [comment]
It looks like I'm going to have to split RelativeIndex into two functions. One to read the index which will use the full path, and the other writes the relative path to the repositories.yaml file

**Comment by rprobotics** [comment]
It looks like I only needed to update init.go everything else was reading the file. 

I was unable to test the 'local' repo since I am unsure how to best do that.

**Comment by Stelminator** [comment]
I've got something I think works for most use cases in https://github.com/rprobotics/helm/pull/1

We're getting close to a full solution.

**Comment by Stelminator** [comment]
@rprobotics I should have fixed the email addresses on my commits, but I think you'll need to do some cleanup on your branch to make the bot happy.

**Comment by Stelminator** [comment]
@bacongobbler Is there anything left preventing this from being merged?

**Comment by Stelminator** [comment]
ah, I think I see: the preference is to no longer write absolute paths at all?

**Comment by bacongobbler** [comment]
Yes.

**Comment by Stelminator** [comment]
I've made an additional change in https://github.com/rprobotics/helm/pull/2 which I think should make things work.  It's not exactly pretty, but please tell me what you think, @rprobotics @bacongobbler

**Comment by rprobotics** [comment]
@Stelminator I'll review ASAP

**Comment by Stelminator** [comment]
@rprobotics @bacongobbler ping

**Comment by rprobotics** [comment]
@Stelminator It seems this PR is blocked because of the 'Changes requested', but I have already taken care of those issues IIRC

**Comment by bacongobbler** [comment]
Speaking for the people over at Microsoft, we're just incredibly busy with other projects at this time and haven't had the time to come back and take a look at this yet. Sorry! We will get back to this!

If you can at least demonstrate that https://github.com/kubernetes/helm/pull/2862#issuecomment-326370130 was addressed with this fix then I might take a closer look. We're planning on releasing Helm v2.8.0 around Kubernetes's 1.9 release around KubeCon so there's no rush/deadline to get this finished.

**Comment by rprobotics** [comment]
I'll try and get to wrapping this up completely this weekend if time permits. Thanks everybody for helping out!

**Comment by rprobotics** [comment]
This week has started really busy for me since end of quarter is coming up. I'll be getting to this ASAP. Sorry for delay

**Comment by rprobotics** [comment]
`rprobotics@C02V32HLHTDF ~> rm -rf .helm/
rprobotics@C02V32HLHTDF ~> helm init
Creating /Users/rprobotics/.helm
Creating /Users/rprobotics/.helm/repository
Creating /Users/rprobotics/.helm/repository/cache
Creating /Users/rprobotics/.helm/repository/local
Creating /Users/rprobotics/.helm/plugins
Creating /Users/rprobotics/.helm/starters
Creating /Users/rprobotics/.helm/cache/archive
Creating /Users/rprobotics/.helm/repository/repositories.yaml
Adding stable repo with URL: https://kubernetes-charts.storage.googleapis.com
Adding local repo with URL: http://127.0.0.1:8879/charts
$HELM_HOME has been configured at /Users/rprobotics/.helm.
Warning: Tiller is already installed in the cluster.
(Use --client-only to suppress this message, or --upgrade to upgrade Tiller to the current version.)
Happy Helming!
rprobotics@C02V32HLHTDF ~> cat ~/.helm/repository/repositories.yaml | grep index.yaml
  cache: stable-index.yaml
  cache: local-index.yaml
rprobotics@C02V32HLHTDF ~> mv .helm/ /tmp/helm
rprobotics@C02V32HLHTDF ~> env HELM_HOME=/tmp/helm/ helm repo update
Hang tight while we grab the latest from your chart repositories...
...Skip local chart repository
...Successfully got an update from the "stable" chart repository
Update Complete. ⎈ Happy Helming!⎈
rprobotics@C02V32HLHTDF ~> env HELM_HOME=/tmp/helm/ helm install stable/cockroachdb
NAME:   inky-grasshopper
LAST DEPLOYED: Sun Dec  3 22:49:59 2017
NAMESPACE: default
STATUS: DEPLOYED

RESOURCES:
==> v1/Service
NAME                                 CLUSTER-IP  EXTERNAL-IP  PORT(S)             AGE
inky-grasshopper-cockroachdb         None        <none>       26257/TCP,8080/TCP  0s
inky-grasshopper-cockroachdb-public  10.0.0.44   <none>       26257/TCP,8080/TCP  0s

==> v1beta1/StatefulSet
NAME                          DESIRED  CURRENT  AGE
inky-grasshopper-cockroachdb  3        1        0s

==> v1beta1/PodDisruptionBudget
NAME                                 MIN-AVAILABLE  MAX-UNAVAILABLE  ALLOWED-DISRUPTIONS  AGE
inky-grasshopper-cockroachdb-budget  67%            N/A              0                    0s
`

**Comment by rprobotics** [comment]
@bacongobbler I reran the tests you mentioned in [#2862 (comment)](url) and posted the results above. The unit-tests are also passing.

**Comment by Stelminator** [comment]
@bacongobbler This diff is rather small, is there more needed for this to be accepted?

**Comment by bacongobbler** [comment]
I'll give this a spin after the holidays. I want to make sure that this change is backwards compatible. Thanks!

**Comment by Stelminator** [comment]
cool, thanks.  I believe it is backward compatible; note that the cache path (e.g. `stable-index.yaml`) is relative to the cache directory, not the root, but that's correct as long as the lines below remain.  If there was a time when cache paths were generated relative to the root, they're already broken via the below.  I think they're currently generated relative to the working directory if not absolute, along with getting the wrong prefix, which is even worse.

https://github.com/kubernetes/helm/blob/8cf72adf521311225274fa835a2b04060eeb781a/pkg/repo/chartrepo.go#L136-L138

**Comment by Stelminator** [comment]
@bacongobbler Could you advise on the right way to merge/rebase/refactor this after https://github.com/kubernetes/helm/pull/3233 was merged?

**Comment by Stelminator** [comment]
@bacongobbler ping.

**Comment by Stelminator** [comment]
@bacongobbler @rprobotics I've rebased and added a couple of commits at https://github.com/Stelminator/helm/tree/paths-2018-01-09

I had a bit of a misunderstanding about `filepath.Rel` involving what it actually does with `base` and which things I should have been comparing, and that has now been remedied.  I've done my best to get imports and unix/windows stuff to be correct, style/etc. comments welcome.

File: cmd/helm/init.go [hunk 1]
```diff
@@ -390,7 +389,12 @@ func initLocalRepo(indexFile, cacheFile string, out io.Writer) (*repo.Entry, err
 		}
 
 		//TODO: take this out and replace with helm update functionality
-		os.Symlink(indexFile, cacheFile)
+		fp, err := filepath.Rel(filepath.Join(home.Cache(), cacheFile), indexFile)
+		if err != nil {
+			return nil, err
+		}
+		pth := home.Path(fp)
+		os.Symlink(pth, filepath.Join(home.Cache(), cacheFile))
```
Comment (pietromenna) [inline cmd/helm/init.go]: Hi,

os.Symlink does not work on Windows. Could you update to use createLink which calls the correct one depending on the OS?

Comment (Stelminator) [reply cmd/helm/init.go]: @pietromenna please find those changes already in my updated pr: https://github.com/kubernetes/helm/pull/3327, specifically the latest commit which adds an error return to createLink: https://github.com/kubernetes/helm/pull/3327/commits/15aee5d8f81c86e9b4869f061dd98e655a5442b8

Comment (pietromenna) [reply cmd/helm/init.go]: Cool! Thanks!

Comment (rprobotics) [reply cmd/helm/init.go]: @Stelminator Should I go ahead and close this PR in favor of yours? I don't want to two PRs open that are trying to solve same issue :)

Comment (Stelminator) [reply cmd/helm/init.go]: @rprobotics If you're ok with my PR, then, yes, please.  I think it's easier to maintain rebased w/o all the merge commits.

**Comment by Stelminator** [comment]
@bacongobbler ping.

**Comment by thomastaylor312** [comment]
I missed that this also introduces a new feature, I have added the feature label as well and put it into 2.9

**Comment by rprobotics** [comment]
@thomastaylor312 Are you ok with me closing this PR in favor of: https://github.com/kubernetes/helm/pull/3327 

?

**Comment by bacongobbler** [comment]
Yes that’s fine. :)

**Pull request closed without merging**